main/src/ca/uwaterloo/flix/language/phase/unification/set/SetUnification.scala

```diff
@@ -0,0 +1,235 @@
+    /** Mark that progress has been made. */
+    def progress(): Unit = somethingHappened = true
+
+    /** Returns `true` if there has been progress. */
+    def hasProgressed: Boolean = somethingHappened
+  }
+
+  object Progress {
+    /** Make a new [[Progress]] object with progress set to `false`. */
+    def apply(): Progress = new Progress()
+  }
+
+  /** A collection of rules that can be used to solve individual [[Equation]]. */
+  private object Rules {
+
+    /**
+      * Solves equations that trivially hold (like `univ ~ univ`) and marks trivially unsolvable
+      * equations (like `univ ~ empty`). Always returns an empty substitution.
+      *
+      * Otherwise returns `eq` directly.
+      *
+      *   - `univ ~ univ` becomes `{}`
+      *   - `univ ~ empty` becomes `{univ !~ empty}`
+      *   - `x1 ~ x1` becomes `{}`
+      *   - `x2 ~ univ` becomes `{x2 ~ univ}`
+      */
+    def trivial(eq: Equation)(implicit p: Progress): (List[Equation], SetSubstitution) = {
+      val Equation(t1, t2, _, _) = eq
+
+      @inline
+      def error(): (List[Equation], SetSubstitution) = {
+        p.progress()
+        (List(eq.toUnsolvable), SetSubstitution.empty)
+      }
+
+      @inline
+      def success(): (List[Equation], SetSubstitution) = {
+        p.progress()
+        (Nil, SetSubstitution.empty)
+      }
+
+      (t1, t2) match {
+        // Equations that are solved.
+        case (Univ, Univ) => success()
+        case (Empty, Empty) => success()
+        case (Cst(c1), Cst(c2)) if c1 == c2 => success()
+        case (Var(x1), Var(x2)) if x1 == x2 => success()
+        case (ElemSet(e1), ElemSet(e2)) if e1 == e2 => success()
+        // Equations that are Unsolvable.
+        case (Univ, Empty) => error()
+        case (Univ, ElemSet(_)) => error()
+        case (Univ, Cst(_)) => error()
+        case (Empty, Univ) => error()
+        case (Empty, ElemSet(_)) => error()
+        case (Empty, Cst(_)) => error()
+        case (ElemSet(_), Univ) => error()
+        case (ElemSet(e), Empty) if e.nonEmpty => error()
+        case (ElemSet(i1), ElemSet(i2)) if i1 != i2 => error()
+        case (ElemSet(_), Cst(_)) => error()
+        case (Cst(_), Univ) => error()
+        case (Cst(_), Empty) => error()
+        case (Cst(_), ElemSet(_)) => error()
+        case (Cst(c1), Cst(c2)) if c1 != c2 => error()
+        // Equations that are not trivial.
+        case _ => (List(eq), SetSubstitution.empty) // Can't do anything.
+      }
+    }
+
+    /**
+      * Solves equations of ground assignments to variables (e.g. `x ~ c1 ∪ e2`).
+      *
+      *   - `x ~ t` where [[SetFormula.isGround]] on `t` becomes `({}, [x -> t])`
+      *   - `!x ~ t` where [[SetFormula.isGround]] on `t` becomes `({}, [x -> !t])`
+      *   - `t1 ∩ t2 ∩ .. ~ univ` becomes `({t1 ~ univ, t2 ~ univ, ..}, [])`
+      *   - `t1 ∪ t2 ∪ .. ~ univ` becomes `({t1 ~ empty, t2 ~ empty, ..}, [])`
+      */
```

Comment:
Do we need some comment about symmetric cases being absent because we assume something of Eq.mk?

Reply:
hmm yeah. And the `!x ~ f` case need to be mirrored. I've though about not having `Equation` ordering invariants. It makes the code hard to read because you have to think about them. Wouldn't it be nicer just to have the mirrored cases? fx `trivial` right now has all cases, disregarding the `Equation` invariants and that seems easier to read to deleting some of lines because of invariants in another file

Reply:
I've tried to rewrite `constantAssignment` to be selfcontained without duplicate cases. Then we could remove `Eq.mk`. What do you think?

Reply:
We could do both:

We still order (for our own sake in debugging)
But the code handles mirrored cases.

Reply:
is ordering good for debugging? to have small left sides?

sounds good

Reply:
Its good to have variables on the left, yes